bon-macros/src/util/mod.rs

```diff
@@ -45,6 +49,52 @@ where
     Ok((params, item))
 }
 
+pub(crate) fn parse_map_macro_input(
+    input: ParseStream<'_>,
+) -> Result<Punctuated<(Expr, Expr), Token![,]>, syn::Error> {
+    Punctuated::parse_terminated_with(input, parse_map_pair)
+}
+
+fn parse_map_pair(pair: ParseStream<'_>) -> Result<(Expr, Expr), syn::Error> {
+    let key = pair.parse().map_err(|_| pair.error(pair.to_string()))?;
+    let _: Token![:] = pair.parse()?;
+    let value = pair.parse()?;
+
+    Ok((key, value))
+}
+
+pub(crate) fn ensure_unique<'k, I>(items: I) -> Result<()>
+where
+    I: IntoIterator<Item = &'k Expr>,
+{
+    let mut errors = Error::accumulator();
+
+    let mut exprs = std::collections::HashSet::new();
+
+    items
+        .into_iter()
+        .filter(|item| is_pure(item))
+        .for_each(|key| {
+            if !exprs.insert(key.clone()) {
+                errors.push(err!(key, "duplicate map key"));
+            }
+        });
+
+    errors.finish()?;
+
+    Ok(())
+}
+
+fn is_pure(expr: &Expr) -> bool {
+    match expr {
+        Expr::Binary(binary) => is_pure(&binary.left) && is_pure(&binary.right),
+        Expr::Lit(_) => true,
+        Expr::Paren(paren) => is_pure(&paren.expr),
```

Comment:
There is also [`Expr::Group`](https://docs.rs/syn/latest/syn/token/struct.Group.html), which is similar to `Paren`, but it's a bit different beast. It is an expression delimited by "invisible" parens. This token can only be created from macro invocations, but we may include its handling here as well similar to `Expr::Parent` logic